src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -143,19 +144,14 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
 
         internal static async ValueTask<TResult> InvokeStreamingServiceAsync<TResult>(
             TService service,
-            Func<TService, Stream, CancellationToken, ValueTask> invocation,
-            Func<Stream, CancellationToken, ValueTask<TResult>> reader,
+            Func<TService, PipeWriter, CancellationToken, ValueTask> invocation,
+            Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
-            // The reader should close the client stream, the writer will close the server stream.
-            // See https://github.com/microsoft/vs-streamjsonrpc/blob/master/doc/oob_streams.md
-            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
-
-            // Create new tasks that both start executing, rather than invoking the delegates directly.
-            // If the reader started synchronously reading before the writer task started it would be blocking, and vice versa
-            // if the writer synchronously filled the buffer before the reader task started it would also be blocking.
-            var writerTask = Task.Run(async () => await invocation(service, serverStream, cancellationToken).ConfigureAwait(false), cancellationToken);
-            var readerTask = Task.Run(async () => await reader(clientStream, cancellationToken).ConfigureAwait(false), cancellationToken);
+            var pipe = new Pipe();
+
+            var writerTask = invocation(service, pipe.Writer, cancellationToken).AsTask();
+            var readerTask = reader(pipe.Reader, cancellationToken).AsTask();
```

Comment:
interesting.  why does the original comment not hold.  i.e.:

```
            // If the reader started synchronously reading before the writer task started it would be blocking, and vice versa
            // if the writer synchronously filled the buffer before the reader task started it would also be blocking.
```

?

Reply:
Both are now expected to only read/write asynchronously. They should never block.

Reply:
i guess my concern is that "being async" != "non-blocking"

i.e. i could totally write something with awaits and wahtnot, and i still may end up doing an initial read/write or something that could end up still blocking.  it seems like that would be very easy to slip in here.  Would it make sense to still just kick off onto the threadpool to ensure that even if it's async+some-blocking, that this will still be ok?

Reply:
Yes, actually this is still be a problem. Need to revert this change. The reader is blocking on the local pipe.

Reply:
Yes, i'll revert this.

Reply:
yaay, i'm not crazy :)